### PR TITLE
refactor(chain)!: Update KeychainTxOutIndex methods to use owned K and ScriptBuf

### LINKED ISSUE #1482: KeychainTxOutIndex API inconsistent with respect to references (should be owned)
Methods often take a reference to a `K` or an owned `K`. They often return a reference or sometimes an owned `K`. I think we should always just used owned. Also I think we should owned `ScriptBuf`'s rather than `Script`. The reason is they are both usually small and fine to copy:

- `K` is usually small and needs to be anyway if you derive lots of spks from a keychain (it is copied each time you do that into the underlying `SpkTxOutIndex`.
- script pubkeys are usually small unless you are doing stuff like DoSing the mempool/utxo set which we don't need to support.

It should be owned everywhere and this should be brought up to the `Wallet` API.

### PR DESCRIPTION
**Pull request by rustaceanrob**
### Description

Make all method signatures of `KeychainTxOutIndex` take owned `K` and use `ScriptBuf` instead of its borrowed counterpart `&Script`. Fixes #1482 

### Notes to the reviewers

Steve also added a CI fix as well

### Changelog notice

- Make all method signatures of `KeychainTxOutIndex` take owned `K`
- Update `KeychainTxOutIndex` methods to use `ScriptBuf`

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

**Comment by notmandatory** [comment]
I pushed commit to also update `KeychainTxOutIndex` to use owned `ScriptBuf`s instead of `&Script`s.

**Comment by notmandatory** [comment]
⛏️ @rustaceanrob one small thing, can you change your commit description to be a little more descriptive, something like:

```
refactor(chain)!: update KeychainTxOutIndex methods to use owned K
```

File: crates/chain/src/indexer/keychain_txout.rs
Comment (storopoli) [inline crates/chain/src/indexer/keychain_txout.rs]: Is this just removing an additional `&` since the `.scan_txout(I) -> &I` (returns a reference) and `keychain` here is already a reference?

File: crates/chain/src/indexer/keychain_txout.rs
Comment (storopoli) [inline crates/chain/src/indexer/keychain_txout.rs:255]: What happens if we enforce that `K` is also `Copy`.

Ref https://github.com/bitcoindevkit/bdk/blob/37cef1da7834d6d29f91c382ae841baf66764c51/crates/chain/src/indexer/keychain_txout.rs#L137

I see from the example in the docstrings that probably `K` can be trait-bounded to `Copy` and then we don't need all those `.clone()`

Comment (notmandatory) [reply crates/chain/src/indexer/keychain_txout.rs:255]: Limiting K to have to implement `Copy` I think would be too restrictive. For instance we wouldn't be able to use types like `String` as K or as fields of a K.

Comment (storopoli) [reply crates/chain/src/indexer/keychain_txout.rs:255]: Yes, then `clone()` is fine.

File: crates/chain/src/indexer/keychain_txout.rs
Comment (rustaceanrob) [reply crates/chain/src/indexer/keychain_txout.rs]: You are right I will fix that

Comment (notmandatory) [reply crates/chain/src/indexer/keychain_txout.rs]: A little higher up in the code (line 142) we've `cloned()` the result of `scan_txout(tx)` so by line 150 it's already an owned K. 
 
But bigger picture, since `replenish_inner_index()` is an internal API an owned K maybe isn't strictly needed here, but since we're making the assumption K is always small I think it's OK to do for consistency.

Comment (notmandatory) [reply crates/chain/src/indexer/keychain_txout.rs]: Oops, I see @rustaceanrob took the other approach of changing `replenish_inner_index()` I'm fine with going that way too.

Comment (rustaceanrob) [reply crates/chain/src/indexer/keychain_txout.rs]: Oh shoot I already changed the inner functions to use a `&K` if there was unneeded clones in my rebase. I agree that the consistency is a nice to have, but since the `replenish_inner_index` logic probably won't change I feel like we might as well take a reference?

**Comment by notmandatory** [comment]
I added a commit to limit tests to 2 threads, this is what I've always had to do locally to fix an issue with `bitcoind` based tests making calls to quickly.

File: crates/chain/src/indexer/keychain_txout.rs
Comment (notmandatory) [reply crates/chain/src/indexer/keychain_txout.rs]: Yup it's fine that you changed `replenish_inner_index` to take a ref.

**Comment by storopoli** [comment]
> I added a commit to limit tests to 2 threads, this is what I've always had to do locally to fix an issue with `bitcoind` based tests making calls to quickly.

Oh the nasty `-- --test-threads=2` strikes back.
Yeah I had tons of issues with that in the past while prepping the Nix CI PR.

**Review by notmandatory — APPROVED** [review]
ACK ade5999bf551bc47fded49ec5c2376073d8d4d5c

We should hold off on merging this until after PR for #1103 is merged so we don't disrupt that work. But may also require a re-base for this PR.

File: crates/chain/src/indexer/keychain_txout.rs
Comment (ValuedMammal) [reply crates/chain/src/indexer/keychain_txout.rs]: I think it's fine to pass in an owned value of `K` since it's cloned in the function body anyway

**Review by ValuedMammal — APPROVED** [review]
ACK ade5999bf551bc47fded49ec5c2376073d8d4d5c

**Comment by notmandatory** [comment]
This needs to be rebased after #1514 is merged.

**Comment by notmandatory** [comment]
I pushed a test rebase, will do final rebase after #1514 is merged.

**Pull request merged**
